### PR TITLE
Implements docker firewall validation

### PR DESCRIPTION
**Pull request by b10n1k**
I added a small validation for the expected configuration of the firewall.
Briefly, makes sure that Docker firewall zone exists with an active docker0
interface and in turn this is not added on trusted zone. Run a container and
verify the iptables.

Applies firewall test in docker and docker_image
the tests implemented based on https://docs.docker.com/network/iptables/

- Related ticket: https://progress.opensuse.org/issues/89872
- Verification run: 
[OSD](https://openqa.suse.de/tests/overview?distri=sle&version=15-SP3&build=b10n1k%2Fos-autoinst-distri-opensuse%2312448)
[TW](https://openqa.opensuse.org/tests/overview?distri=opensuse&build=b10n1k%2Fos-autoinst-distri-opensuse%2312448&version=Tumbleweed)
[Leap](https://openqa.opensuse.org/tests/overview?distri=opensuse&build=b10n1k%2Fos-autoinst-distri-opensuse%2312448&version=15.3)
[sle-12-SP5](https://openqa.suse.de/tests/5954410)

### CODE REVIEW

**Comment by b10n1k** [comment]
@jlausuch https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/12448#discussion_r625724050

File: lib/containers/utils.pm [hunk 1]
```diff
@@ -41,20 +42,62 @@ sub test_seccomp {
...
+    my ($container, $runtime) = @_;
+    my $ip = script_output "$runtime inspect $container --format='{{.NetworkSettings.IPAddress}}'";
+    record_info "$ip";
+    return $ip;
+}
+
+sub registry_url {
+    my ($container_name, $version_tag) = @_;
     my $registry = get_var('REGISTRY', 'docker.io');
+    $registry = trim($registry);
```
Comment (pdostal) [inline lib/containers/utils.pm:84]: This is nice idea, but `trim()` does not cover the trailing slash at the end:
```suggestion
    my $registry = trim(get_var('REGISTRY', 'docker.io'));
    $registry =~ s{/$}{};
```

**Review by pdostal — APPROVED** [review]
Awesome Yiannis! Just few last touches and we can merge it today.

File: lib/containers/utils.pm (hunk 1 above)
Comment (b10n1k) [reply lib/containers/utils.pm:84]: https://docs.mojolicious.org/Mojolicious/Validator#trim seems to imply the opposite

File: tests/containers/docker_image.pm [hunk 2]
```diff
@@ -23,22 +24,23 @@ use Mojo::Base qw(consoletest);
...
+    my ($self) = @_;
     my ($image_names, $stable_names) = get_suse_container_urls();
     my ($running_version, $sp, $host_distri) = get_os_release;
     my $runtime = "docker";
 
     install_docker_when_needed($host_distri);
     allow_selected_insecure_registries(runtime => $runtime);
     scc_apply_docker_image_credentials() if (get_var('SCC_DOCKER_IMAGE'));
-
+    check_docker_firewall() if (check_runtime_version(get_docker_version(), ">=20.10") && $self->firewall() eq 'firewalld' && $host_distri ne 'centos' && $host_distri ne 'ubuntu');
```
[the remark this answers is not in the document]
Comment (b10n1k) [reply tests/containers/docker_image.pm]: Done

**Comment by asmorodskyi** [comment]
at least TW VR is pointing to another branch https://openqa.opensuse.org/tests/1728772#settings

File: lib/containers/utils.pm [hunk 3]
```diff
@@ -41,20 +43,62 @@ sub test_seccomp {
...
 }
 
-sub basic_container_tests {
-    my %args    = @_;
-    my $runtime = $args{runtime};
-    die "You must define the runtime!" unless $runtime;
+sub check_docker_firewall {
+    my $container_name = 'sut_container';
+    my $running        = script_output qq(docker ps -a -q | wc -l);
+    validate_script_output('ip a s docker0', sub { /state DOWN/ }) if $running != 0;
```
Comment (asmorodskyi) [inline lib/containers/utils.pm:49]: so `$running`  will be not 0 when at least one container is running , right ? Isn't actually `docker0` interface suppose to be UP in that case ?

File: lib/containers/utils.pm [hunk 4]
```diff
@@ -41,20 +43,62 @@ sub test_seccomp {
     }
 }
 
-sub basic_container_tests {
-    my %args    = @_;
-    my $runtime = $args{runtime};
-    die "You must define the runtime!" unless $runtime;
+sub check_docker_firewall {
```
Comment (asmorodskyi) [inline lib/containers/utils.pm:46]: I would add check if firewall running at all

File: lib/containers/utils.pm [hunk 5]
```diff
@@ -41,20 +43,62 @@ sub test_seccomp {
...
+    my $container_name = 'sut_container';
+    my $running        = script_output qq(docker ps -a -q | wc -l);
+    validate_script_output('ip a s docker0', sub { /state DOWN/ }) if $running != 0;
+    assert_script_run "firewall-cmd --list-all --zone=docker";
+    validate_script_output "firewall-cmd --list-interfaces --zone=docker",  sub { /docker0/ };
+    validate_script_output "firewall-cmd --list-interfaces --zone=trusted", sub { /^\s*$/ };
```
Comment (asmorodskyi) [inline lib/containers/utils.pm:52]: what this line suppose to prove ? that **some** interfaces are in `trusted` ? how it is related to docker + firewalld ?

File: lib/containers/utils.pm [hunk 6]
```diff
@@ -41,20 +43,62 @@ sub test_seccomp {
...
+    assert_script_run "docker kill $container_name ";
+}
+
+sub get_docker_version {
+    my $v = script_output("docker --version");
+    record_info "$v", $v =~ /(\d{2}\.\d{2})/;
+    return $v =~ /(\d{2}\.\d{2})/;
+}
+
+sub check_runtime_version {
```
Comment (asmorodskyi) [inline lib/containers/utils.pm:70]: not sure what purpose of this function . you just proxy'ing two params into `version_utils::check_version` maybe we can just call `version_utils::check_version` where you calling check_runtime_version ?

File: lib/containers/utils.pm [hunk 7]
```diff
@@ -41,20 +43,62 @@ sub test_seccomp {
...
+    my ($current, $other) = @_;
+    return check_version($other, $current, qr/\d{2}(?:\.\d+)/);
+}
 
+sub container_ip {
+    my ($container, $runtime) = @_;
+    my $ip = script_output "$runtime inspect $container --format='{{.NetworkSettings.IPAddress}}'";
+    record_info "$ip";
```
Comment (asmorodskyi) [inline lib/containers/utils.pm:78]:
```suggestion
    record_info "container IP", "$ip";
```
it is good when you know the code , but when you don't box with random IP may confuse ...

File: tests/containers/docker.pm [hunk 8]
```diff
@@ -50,6 +51,7 @@ sub run {
     test_seccomp();
     allow_selected_insecure_registries(runtime => 'docker');
 
+    check_docker_firewall() if (check_runtime_version(get_docker_version(), ">=20.10") && $self->firewall() eq 'firewalld');
```
Comment (asmorodskyi) [inline tests/containers/docker.pm:54]: I think it would be nice to add some explanation why you want to run firewall check in versions greater than `20.10`

File: tests/containers/docker_image.pm [hunk 9]
```diff
@@ -21,8 +21,8 @@
 
 use Mojo::Base qw(consoletest);
 use testapi;
-use utils;
 use containers::common;
+use containers::utils;
```
Comment (asmorodskyi) [inline tests/containers/docker_image.pm:25]: please revert this changes

File: lib/containers/utils.pm (hunk 1 above)
Comment (pdostal) [reply lib/containers/utils.pm:85]: > https://docs.mojolicious.org/Mojolicious/Validator#trim seems to imply the opposite

I do not agree.

```
$ vim /tmp/trim.pl
use Mojo::Util 'trim';
my $str = "hub.docker.io:5000/";
print($str."\n");
$str = trim($str);
print($str."\n");
$str =~ s{/$}{};
print($str."\n");
$ perl /tmp/trim
hub.docker.io:5000/
hub.docker.io:5000/
hub.docker.io:5000
```
